src/PowerShellEditorServices/Server/PsesServiceCollectionExtensions.cs

```diff
@@ -24,16 +24,21 @@ internal static class PsesServiceCollectionExtensions
                     (provider) =>
                         PowerShellContextService.Create(
                             provider.GetService<ILoggerFactory>(),
+                            // NOTE: Giving the context service access to the language server this
+                            // early is dangerous because it allows it to start sending
+                            // notifications etc. before it has initialized, potentially resulting
+                            // in deadlocks. We're working on a solution to this.
                             provider.GetService<OmniSharp.Extensions.LanguageServer.Protocol.Server.ILanguageServerFacade>(),
                             hostStartupInfo))
-                .AddSingleton<TemplateService>()
+                .AddSingleton<TemplateService>() // TODO: What's the difference between this and the TemplateHandler?
```

Comment:
Handlers handle requests from the LSP client, like ASP.NET controllers.

Services internally manage some conceptual area, and generally implement the business logic behind handlers.

Reply:
Helpful!